module_3/Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java

```diff
@@ -0,0 +1,126 @@
+package com.epam.ems.dao.certificatedao;
+
+import com.epam.ems.dao.CRDDao;
+import com.epam.ems.dao.CRUDDao;
+import com.epam.ems.dto.Certificate;
+import com.epam.ems.dto.Tag;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.criteria.*;
+import javax.transaction.Transactional;
+import java.util.Collection;
+import java.util.List;
+
+import static com.epam.ems.dto.fields.Constants.*;
+
+
+@Repository
+public class CertificateDaoImpl implements CRUDDao<Certificate> {
+
+    private EntityManager entityManager;
+
+    private CriteriaBuilder criteriaBuilder;
```

Comment:
This is not thread-safe - your class is a singleton.
Think what happens when several requests would process class variable?